src/Database/LSMTree/Internal/TableHandle.hs

```diff
@@ -0,0 +1,183 @@
+  , sessionUniqCounter :: !(StrictMVar m Integer)
+  }
+
+{-------------------------------------------------------------------------------
+  Table handle
+-------------------------------------------------------------------------------}
+
+-- | A handle to an on-disk key\/value table.
+--
+-- For more information, see 'Database.LSMTree.Normal.TableHandle'.
+data TableHandle m h = TableHandle {
+    tableSession     :: !(Session m h)
+  , tableConfig      :: !TableConfig
+  , tableWriteBuffer :: !(WriteBuffer)
+    -- | A hierarchy of levels. The vector indexes double as level numbers.
+  , tableLevels      :: !(V.Vector (Level h))
+    -- | Cache of flattened 'levels'.
+    --
+    -- INVARIANT: when 'level's is modified, this cache should be updated as
+    -- well, for example using 'mkLevelsCache'.
+  , tableLevelsCache :: !(LevelsCache h)
+  }
+
+-- | Runs in order from newer to older
+newtype Level h = Level {
+    residentRuns :: V.Vector (Run h)
+{- TODO: this is where ongoing merges appear once we implement scheduling ,
+  incomingRuns :: MergingRun
+-}
+  }
+
+-- | Flattend cache of the runs that referenced by a table handle.
+--
+-- This cache includes a vector of runs, but also vectors of the runs broken
+-- down into components, like bloom filters, fence pointer indexes and file
+-- handles. This allows for quick access in the lookup code. Recomputing this
+-- cache should be relatively rare.
+--
+-- Use 'mkLevelsCache' to ensure that there are no mismatches between the vector
+-- of runs and the vectors of run components.
+data LevelsCache h = LevelsCache_ {
+    cachedRuns      :: !(V.Vector (Run h))
+  , cachedFilters   :: !(V.Vector (Bloom SerialisedKey))
+  , cachedIndexes   :: !(V.Vector IndexCompact)
+  , cachedKOpsFiles :: !(V.Vector h)
+  }
+
+-- | Flatten the argument 'Level's into a single vector of runs, and use that to
+-- populate the 'LevelsCache'.
+mkLevelsCache :: V.Vector (Level h) -> LevelsCache h
+mkLevelsCache lvls = LevelsCache_ {
+      cachedRuns      = rs
+    , cachedFilters   = V.map Run.runFilter rs
+    , cachedIndexes   = V.map Run.runIndex rs
+    , cachedKOpsFiles = V.map Run.runKOpsFile rs
+    }
+  where rs = V.concatMap residentRuns lvls
+
+{-------------------------------------------------------------------------------
+  Configuration
+-------------------------------------------------------------------------------}
+
+-- | Table configuration parameters, including LSM tree tuning parameters.
+--
+-- Some config options are fixed (for now):
+--
+-- * Merge policy: Tiering
+--
+-- * Size ratio: 4
+data TableConfig = TableConfig {
+    confMergePolicy      :: !MergePolicy
+    -- Size ratio between the capacities of adjacent levels.
+  , confSizeRatio        :: !SizeRatio
+    -- | Total number of bytes that the write buffer can use.
+    --
+    -- The maximum is 4GiB, which should be more than enough for realistic
+    -- applications.
+  , confWriteBufferAlloc :: !Word32
```

Comment:
Do we already have a clear idea of what the buffer flush policy is going to be more precisely?
The buffer gets flushed once it reaches the specified size in memory (i.e. serialised form of keys/values, including the `Entry` and `Data.Map` overhead)? Does it strictly need to be flushed before getting any larger or can a batch of updates that crosses the size still be completely inserted before flushing?

Reply:
My first idea would be to base it purely on the serialised size of keys and values, but that's not set in stone. We can try out different approaches when start implementing this

Reply:
It depends a bit on whether the config values should be seen as strict upper bounds, but in most cases that shouldn't be necessary.